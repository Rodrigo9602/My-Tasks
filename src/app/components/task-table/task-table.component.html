<div class="task-table-message" *ngIf="tasks.length === 0; else taskTableWrapper">
    <p class="header">You have not task registered yet!</p>
    <p class="subheader">Lets change that</p>
    <button class="btn-primary" (click)="onAddTask()">Add new task</button>
</div>




<ng-template #taskTableWrapper>
    <div class="task-table-wrapper">
        <header>
            <p>{{userName}}'s tasks</p>
            <button class="btn-primary" (click)="onAddTask()">Add new task</button>
        </header>

        <div class="table-wrapper">
            <table>
                <thead>
                    <tr>
                        <th>name</th>
                        <th>description</th>
                        <th>creation date</th>
                        <th>ending date</th>
                        <th>state</th>
                        <th>options</th>
                    </tr>
                </thead>

                <tbody>
                    <tr *ngFor="let task of pagedTasks">
                        <td>{{task.name}}</td>
                        <td>{{task.description}}</td>
                        <td>{{formatDate(task.creationDate)}}</td>
                        <td>{{formatDate(task.endingDate)}}</td>
                        <td [ngSwitch]="task.state">
                            <div class="square" *ngSwitchCase="'creado'" style="background-color: #1f86f3;"></div>
                            <div class="square" *ngSwitchCase="'en proceso'" style="background-color: #eff31f;"></div>
                            <div class="square" *ngSwitchCase="'terminado'" style="background-color: #58f31f;"></div>
                            <div class="square" *ngSwitchCase="'no completado'" style="background-color: #f31f1f;">
                            </div>
                        </td>
                        <td>
                            <div class="options-wrapper">
                                <button (click)="updateTask(task)" class="btn-icon">
                                    <fa-icon [icon]="updateIcon"></fa-icon>
                                </button>
                                <button (click)="deleteTask(task)" class="btn-icon">
                                    <fa-icon [icon]="deleteIcon"></fa-icon>
                                </button>
                            </div>
                        </td>
                    </tr>
                </tbody>
            </table>

            <mat-paginator [length]="tasks.length" [pageSize]="pageSize"
                (page)="handlePageEvent($event)" aria-label="Select page">
            </mat-paginator>
        </div>
    </div>
</ng-template>