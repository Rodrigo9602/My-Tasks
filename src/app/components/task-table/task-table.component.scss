.task-table-message {
    width: 72vw;    
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    gap: 0;

    p { 
        font-family: 'poppins';        
        color: var(--text-color);
    }

    p:nth-child(2){         
        margin-top: -30px;
    }
}




.task-table-wrapper {
    width: 100%;    
    padding-top: 20px;
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center; 
    gap: 20px;   

    header {
        width: 90%;
        height: 6vh;            
        display: flex;            
        align-items: center;
        justify-content: space-between;
        padding: 0 30px;

        p {
            font-family: 'poppins';
            color: var(--text-color);            
        }
    }    


    .table-wrapper {
        width: 100%;        
        display: flex;  
        flex-direction: column;          
        align-items: center;
        justify-content: space-between;

        table {
            width: 90%;
            margin: 0.5em auto;
            font-family: 'kabelC';
            border-collapse: collapse;
            border: 1px solid var(--text-color);    
        }
        
        thead {    
            height: 3em;
            pointer-events: none;
        }
        
        tr {
            text-align: center;    
            transition: all 0.3s ease;
        }
        
        th {   
            font-size: 18px; 
            text-align: center;
            background-color: var(--items-color-gradient-sec);
            color: #fff;
        }
        
        td {
            font-size: 16px;
            padding: 10px;
            width: 300px; 
            color: var(--text-color);             
            
            .options-wrapper {
                width: 100%;    
                display: flex;
                align-items: center;
                justify-content: center;
                gap: 15px;    
            }

            .square {
                width: 1rem;
                height: 1rem;
                border-radius: 0.2rem;
                margin: 0 auto;
            }
        }

        mat-paginator {
            background-color: var(--bg);
            color: var(--text-color);
        }
        
        
    }
}



