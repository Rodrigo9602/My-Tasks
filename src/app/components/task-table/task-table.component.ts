import { Component, Input, OnDestroy, OnInit, Output, EventEmitter } from '@angular/core';
import { CommonModule, DatePipe } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { FontAwesomeModule } from '@fortawesome/angular-fontawesome';
import { MatDialogModule } from '@angular/material/dialog';
import {MatPaginatorModule} from '@angular/material/paginator';
import { MatDialog } from '@angular/material/dialog';

import { DialogComponent } from '../dialog/dialog.component';

import { TaskService } from '../../services/task.service';

import { Subscription } from 'rxjs';
import { Task } from '../../models/task.model';


import { faPenToSquare, faTrashAlt } from '@fortawesome/free-regular-svg-icons';
import { DataService } from '../../services/data.service';
import { Toast } from '../../global/toast.global';

@Component({
  selector: 'app-task-table',
  standalone: true,
  imports: [CommonModule, FormsModule, FontAwesomeModule, MatDialogModule, MatPaginatorModule],
  providers: [DatePipe],
  templateUrl: './task-table.component.html',
  styleUrl: './task-table.component.scss'
})
export class TaskTableComponent implements OnInit, OnDestroy {
  public tasks: Array<Task> = [];  

  // pagination vars
  public pagedTasks: Array<Task> = [];
  public pageSize:number = 5;
  public pageIndex:number = 0;


  public deleteIcon = faTrashAlt;
  public updateIcon = faPenToSquare;
  
  @Input() userName: string = '';
  @Output() taskAdded = new EventEmitter<Task>();
  @Output() taskRemoved = new EventEmitter<Task>();

  private suscription:Subscription | undefined;

  constructor(private _dialog: MatDialog, private _taskService:TaskService, public datePipe: DatePipe, private dataService: DataService) { }

  ngOnInit(): void { 
    this.dataService.task$.subscribe(tasks => {      
      this.tasks = tasks;

      this.handlePageEvent({
        pageIndex: 0,
        pageSize: this.pageSize,  
      });
    });
  }

  onAddTask() {
    const dialogRef = this._dialog.open(
      DialogComponent, {
      width: '30%',
      minWidth: '25rem',
      height: '500px',
      data: { form: 'addTask', dataObject: {} }
    });

    dialogRef.afterClosed().subscribe(res => {
      if (res !== undefined) {        
        this.suscription = this._taskService.save(res.name, res.description, res.endingDate).subscribe({
          next: res => {            
            this.taskAdded.emit(res);
          },
          error: e => {
            Toast.fire({
              icon: 'error',
              title: e.error.message
            });
          }
        });
      }
    });
  }


  handlePageEvent(event:any) {
    this.pageIndex = event.pageIndex;
    this.pageSize = event.pageSize;    
    this.pagedTasks = this.tasks.slice(this.pageIndex * this.pageSize, this.pageIndex * this.pageSize + this.pageSize);
    
  }

  formatDate(dateIso: Date) { 
    const date = new Date(dateIso);   
    const localDate = new Date(date.getTime() + date.getTimezoneOffset()*60*1000);
    return this.datePipe.transform(localDate, 'dd/MM/yyyy');
  }

  updateTask(task:Task) {

  }

  deleteTask(task:Task) {
    this._taskService.delete(task._id).subscribe({
      next: res => {       
        this.taskRemoved.emit(task);
      },
      error: e => {
        console.log(e);
      }
    });
  }

  ngOnDestroy(): void {
    this.suscription?.unsubscribe();
  }
}
