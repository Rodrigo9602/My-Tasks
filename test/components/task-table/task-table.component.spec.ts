import { ComponentFixture, TestBed } from '@angular/core/testing';

import { TaskTableComponent } from '../../../src/app/components/task-table/task-table.component';
import { HttpClientTestingModule } from '@angular/common/http/testing';


describe('TaskTableComponent', () => {
  let component: TaskTableComponent;
  let fixture: ComponentFixture<TaskTableComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      imports: [TaskTableComponent, HttpClientTestingModule],      
    })
    .compileComponents();
    
    fixture = TestBed.createComponent(TaskTableComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match initial snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });
});
