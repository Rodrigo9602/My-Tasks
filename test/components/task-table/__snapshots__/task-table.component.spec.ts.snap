// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskTableComponent should match initial snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="task-table-message"
  >
    <p
      class="header"
    >
      You have not task registered yet!
    </p>
    <p
      class="subheader"
    >
      Lets change that
    </p>
    <button
      class="btn-primary"
    >
      Add new task
    </button>
  </div>
  
  
</div>
`;
